doc/neps/nep-00XX-array-function-protocol.rst

```diff
@@ -0,0 +1,511 @@
+   complexity.
+
+Finding and calling the right ``__array_function__``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Given a Numpy function, ``*args`` and ``**kwargs`` inputs, we need to
+search through ``*args`` and ``**kwargs`` for all appropriate inputs
+that might have the ``__array_function__`` attribute. Then we need to
+select among those possible methods and execute the right one.
+Negotiating between several possible implementations can be complex.
+
+Finding arguments
+'''''''''''''''''
+
+Valid arguments may be directly in the ``*args`` and ``**kwargs``, such
+as in the case for ``np.tensordot(left, right, out=out)``, or they may
+be nested within lists or dictionaries, such as in the case of
+``np.concatenate([x, y, z])``. This can be problematic for two reasons:
+
+1. Some functions are given long lists of values, and traversing them
+   might be prohibitively expensive
+2. Some function may have arguments that we don't want to inspect, even
+   if they have the ``__array_function__`` method
+
+To resolve these we ask the functions to provide an explicit list of
+arguments that should be traversed. This is the ``relevant_arguments=``
+keyword in the examples below.
+
+Trying ``__array_function__`` methods until the right one works
+'''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''
+
+Many arguments may implement the ``__array_function__`` protocol. Some
+of these may decide that, given the available inputs, they are unable to
+determine the correct result. How do we call the right one? If several
+are valid then which has precedence?
+
+The rules for dispatch with ``__array_function__`` match those for
+``__array_ufunc__`` (see
+`NEP-13 <http://www.numpy.org/neps/nep-0013-ufunc-overrides.html>`__).
+In particular:
+
+-  NumPy will gather implementations of ``__array_functions__`` from all
+   specified inputs and call them in order: subclasses before
+   superclasses, and otherwise left to right.
+-  Implementations of ``__array_function__`` indicate that they can
+   handle the operation by returning any value other than
+   ``NotImplemented``.
+-  If all ``__array_function__`` methods return ``NotImplemented``,
+   NumPy will raise ``TypeError``.
+
+Changes within Numpy functions
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Given a function defined above, for now call it
+``do_array_function_dance``, we now need to call that function from
+within every relevant Numpy function. This is a pervasive change, but of
+fairly simple and innocuous code that should complete quickly and
+without effect if no arguments implement the ``__array_function__``
+protocol. Let us consider a few examples of NumPy functions and how they
+might be affected by this change:
+
+.. code:: python
+
+    def broadcast_to(array, shape, subok=False):
+        success, value = do_array_function_dance(
+            func=broadcast_to,
+            relevant_arguments=[array],
+            array, shape, subok=subok)  # *args, **kwargs
```

Comment:
This does not match the function signature `(self, func, overload_types, *args, **kwargs)` and mixes positional and keyword arguments. In particular, I don't understand what is input as the `overload_types` positional argument.